[Joel Ho] iP

JoelHo opened the pull request.

Commit:
Add Gradle support

Commit:
build.gradle: Update version to 8.29

JoelHo committed:
Abstracted DukeBot, Implemented greet and bye

Abstracted DukeBot

Made Duke class solely the entry point to the program
Bot functions abstracted into DukeBot class
Localisations implemented by Messages class

Done so to ensure DukeBot can be instanced, rather than only keep track of 1 state of Duke


Implement basic greet and bye

Duke can now say hello and goodbye before exiting

JoelHo committed:
Merge remote-tracking branch 'nus-cs2103-AY2021S2/master'

Merge upstream README change

JoelHo committed:
Add ability to parse commands, implement echo and exit

Add Parser class to deal with input commands
* Also keeps track of whether the command is to exit for now

Add Echo and Exit, implementing Command interface
* Command is an interface with a single method execute() to unify all commands
* Echo is a placeholder command, to be deprecated in next release

JoelHo committed:
Renamed Messages to DukeStrings

More appropriate name to reflect use of class in future

JoelHo committed:
Fixed duplicate shutdown message

JoelHo committed:
Refactor UI strings, Change echo to add

Refactor strings used for UI for better reference

Rename echo class to add
* No functionality implemented

JoelHo committed:
Add ListCommand and TaskList

Add skeleton classes ListCommand and TaskList
Update parser to accept list

JoelHo committed:
Fix typo in UI strings, Remove extra modifier from interface

JoelHo committed:
Implement add and list commands

Implemented add and list functionality
* Changed method signature for Command.execute to accept taskList

JoelHo committed:
Add Task skeleton

JoelHo committed:
Update Duke to use Task, fix numbering with list

JoelHo committed:
Implement done command

Introduce done command to Parser
Enabled TaskList to mark tasks as done
Updated strings

Replaced undone Task printing unicode ✘ with letter X to maintain fixed-width on some consoles

JoelHo committed:
Added Deadline, Event, Todo task skeletons

JoelHo committed:
Add Todo, Event, Deadline functionality

Add new commands for todo, event, deadline
* Extends add, superclass used to update task list

Add new InvalidCommandException
* Used when the command passed to Duke is not understood

Handle InvalidCommandException in DukeBot

JoelHo committed:
Update List message

Made the list command more human friendly
Default message when there are no tasks

JoelHo committed:
Add basic tests

Add tests using I/O redirection
Modify test scripts to compile with UTF-8 encoding
* Allows ✓ in Task to be recognized

JoelHo committed:
Handle whitespace only

Throw exception when the scanner cannot find any tokens (i.e. only whitespace entered)

JoelHo committed:
Fix whitespace in todo output, Update tests

Fix extra whitespace appearing in todo tasks
* Update tests to reflect this

Add a few more tests to ensure proper rejection of garbage input

JoelHo committed:
Rename DukeStrings > DukeString

JoelHo committed:
Move InvalidCommand inside DukeException

DukeException will collect all exceptions related to Duke

Move InvalidCommand message to DukeString

JoelHo committed:
Add new exceptions

Add new exceptions for invalid task description and invalid task number

JoelHo committed:
Implement InvalidTask and EmptyDescription

Implement InvalidTask and EmptyDescription exceptions
* Check if done is specified with no number/invalid number
* Check that task is not specified with empty description

JoelHo committed:
Fix styling issues across all classes

Minor styling pass, mostly marking parameters as final for safety
Ensure DukeString cannot be instanced/extended

JoelHo committed:
Update tests with new exceptions

Update tests to test for new exceptions implemented

JoelHo committed:
Add delete skeleton, strings

Add skeleton class for delete, strings related to delete

JoelHo committed:
Implement delete

Implement delete command, throwing InvalidTask
* Add deleteTask() to TaskList to support delete operation

JoelHo committed:
Update tests with delete

Insert new tests for delete and invalid delete

JoelHo committed:
Cleanup classes, put into packages

Divide classes into packages before proceeding for ease of future organisation

Reformat Parser to better fit coding standard

JoelHo committed:
Add support for de/serialisation of Task/List

Add de/serialisation of TaskList
* Calls de/serialisation of Task

Add de/serialisation of Task
* Split tokens by \255 (should not be an enter-able character!)

Add new exception for malformed data

JoelHo committed:
Add Storage functionality

Make Duke store tasklist every input loop
* Storage file specified by DukeBot constructor

New exception for write error

Fix bug with case not breaking in TaskList

Fix serialisation of TodoTask

JoelHo committed:
Handle malformed storage

If storage is malformed, make copy of file so the users data is not lost

JoelHo committed:
Initial implementation of LocalDateTime in tasks

Convert Deadline and Event tasks/commands to use LocalDateTime

Move parsing of Deadline and Event times to Parser

Introduce new Exception for Deadline and Event if time is not specified

Take in time in ISO format for initial implementation (YYYY-MM-DDTHH:MM:SS)

JoelHo committed:
Update event to have end, Improve error handling

Update event to have an end date, specified by /to

Improve error handling
* New exceptions for empty deadline date, empty event date and event end before event start

JoelHo committed:
Handle EmptyDeadlineDate

Handle exception previously uncaught

JoelHo committed:
Change format of event and deadline time

Change event and deadline time to more human friendly format

JoelHo committed:
Update tests for new date handling

Update tests to work with the new date handling
* Test blank descriptions and dates more rigorously

JoelHo committed:
Merge branch 'branch-Level-7'

Merge storage functionality to master

JoelHo committed:
Merge branch 'branch-Level-8'

Merge date handling functionality to master

JoelHo committed:
Update de/serialisation of tasks using new date

Made de/serialisation of tasks compatible with the new date format

JoelHo committed:
Abstract Ui into class

Abstract duke output into its own class

Add a special separator when duke encounters an error

JoelHo committed:
Add basic tests

Test de/serialisation of deadline and event

Test parser and tasklist is able to add and delete 1 task

JoelHo committed:
Include autogenerated manifest

JoelHo committed:
Include javadocs for most public methods, Minor style fixes

JoelHo committed:
Update javadocs for some classes

JoelHo committed:
Reformat classes to better suit standard

JoelHo committed:
Added FindCommand class and supporting methods

JoelHo committed:
Add support for find command in Duke

Enable Duke to use the find command

JoelHo committed:
Merge branch 'branch-A-JavaDoc'

JoelHo committed:
Merge branch 'A-CodingStandard'

JoelHo committed:
Merge branch 'Level-9'

JoelHo committed:
Merge branch 'Level-9'

Add support for find command to Duke

JoelHo committed:
Merge remote-tracking branch 'origin/master'

JoelHo committed:
Merge remote-tracking branch 'origin/add-gradle-support' into branch-A-Gradle

JoelHo committed:
Update gradle config

Update gradle config to use UTF-8 and run the correct Main class

JoelHo committed:
Add checkstyle config 

Add checkstyle config from https://github.com/se-edu/addressbook-level3/tree/master/config/checkstyle

JoelHo committed:
Format files to fit style guide

JoelHo committed:
Clean up minor styling errors

Clean up some styling errors

Added descriptive comments for all classes

JoelHo committed:
Merge branch 'branch-A-Gradle'

Add Gradle and checkstyle support

tashawan23 commented on src/main/java/duke/task/DeadlineTask.java:
```diff
@@ -0,0 +1,59 @@
+        this.deadline = date;
+    }
+
+    private DeadlineTask(boolean done, String desc, LocalDateTime date) {
```
Would it be better to change parameter boolean done to isDone?

JoelHo committed:
Update Gradle Configuration

Add JavaFX to gradle config

tashawan23 commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,152 @@
+     * Getter for the size of the list.
+     * @return the number of tasks in the list
+     */
+    public int size() {
```
Maybe it would be better to change the method name to getSize() since it is a getter

tashawan23 commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,152 @@
+     * @param idx the index of the task to be marked as done
+     * @return the String format of the specified task
+     */
+    public String doneTask(final int idx) {
```
Maybe it will be better to change doneTask to a verb such as checkTask

JoelHo replied:
Thanks for your comments, really appreciate it :). Definitely missed a few naming issues here and there!

JoelHo committed:
Move Duke class into package duke

JoelHo committed:
Merge branch 'master' into branch-Level-10

tashawan23 reviewed (commented):
Overall, i found your code easy to understand. It was difficult to find any coding standard violations especially in terms of your layout:) I noted some changes that could be made to method names so that it may be better to understand.

JoelHo committed:
Add JavaFX GUI, Deprecate CLI interface

Add JavaFX GUI to Duke

Moved CLI interface of Duke to DukeCli, now deprecated

Refactored DukeBot to handle input without a scanner

JoelHo committed:
Add javadocs for new methods

JoelHo committed:
Tidy up javadocs

JoelHo committed:
Tidy method names

Tidy method names to better fit coding standard (thanks @tashawan23!)

JoelHo committed:
Update build.gradle

Update build.grade to correctly run Launcher

Fix redundant import from MainWindow

Fix Main not having newline before EOF

JoelHo committed:
Catch edge case with indexed tasks

Catch zero/negative indexed tasks

JoelHo committed:
Merge branch 'branch-Level-10'

JoelHo committed:
Fix styling issues

Fix more styling issues, add javadoc support for UTF-8

seaniy commented on src/main/java/duke/task/TodoTask.java:
```diff
@@ -0,0 +1,48 @@
+        super(desc);
+    }
+
+    private TodoTask(boolean done, String desc) {
```
Would variable name for boolean isDone be better?

seaniy commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,44 @@
+     * @param term the term to search for.
+     * @return a boolean containing whether the description contains the term.
+     */
+    public boolean containsTerm(String term) {
```
Would containsKeyword be a more appropriate method name?

seaniy commented on src/main/java/duke/interaction/Ui.java:
```diff
@@ -0,0 +1,33 @@
+     * @param msg the message to be printed.
+     */
+    public static void printOut(final String msg) {
+        System.out.println(DukeString.SEPARATOR);
```
Perhaps joining the separators as a string before and after the msg and then printing the output would be better instead of calling System.out.println() multiple times?

seaniy commented on src/main/java/duke/common/DukeString.java:
```diff
@@ -0,0 +1,47 @@
+            "Sorry, there has been an error reading your tasks.\nMaking a copy of your tasks at data/duke.bak.";
+    public static final String EXCEPTION_STORAGE_WRITE_ERROR = "Sorry, there has been an error writing your tasks.";
+
+    private DukeString() {
```
Perhaps an empty constructor would be unnecessary to be included as a default constructor can be used instead?

JoelHo replied:
Thanks a lot for your comments, much appreciated! Will take a look at the issues you mentioned, but in this instance the intent of the constructor is so DukeString can't be instantiated (https://rules.sonarsource.com/java/RSPEC-1118) (https://checkstyle.sourceforge.io/config_design.html#HideUtilityClassConstructor). It's not mentioned in the style guide but I thought it would be good to include. Let me know if you disagree still!

seaniy commented on src/main/java/duke/common/DukeException.java:
```diff
@@ -0,0 +1,55 @@
+            super(DukeString.EXCEPTION_EMPTY_DEADLINE_DATE);
+        }
+    }
+
```
Extra blankspace here

seaniy commented on src/main/java/duke/command/ExitCommand.java:
```diff
@@ -0,0 +1,20 @@
+ * Represents a command that tells the parser to exit.
+ */
+public class ExitCommand implements Command {
+    /**
```
Perhaps you could standardize if you would like to add a blank line after declaring class or not? I have noticed inconsistencies of blank spaces after class declaration in several other places too.

seaniy reviewed (commented):
Overall code is well written and meets most of the coding standards. Great job!

JoelHo committed:
Add assertions to ensure code integrity

JoelHo committed:
Fix code quality issues

Fix some quality issues (thanks @seaniy !)

JoelHo committed:
Merge pull request #1 from JoelHo/branch-A-CodeQuality

Fix code quality issues

JoelHo committed:
Merge branch 'master' into branch-A-Assertions

JoelHo committed:
Merge pull request #2 from JoelHo/branch-A-Assertions

Add assertions to ensure code integrity

JoelHo committed:
Add gradle.yml for github actions

JoelHo committed:
Import checks from tp

JoelHo committed:
Change Unix permissions for shell scripts

JoelHo committed:
Change all line endings to LF from CRLF

JoelHo committed:
Add newline at EOF to input redirection tests

JoelHo committed:
Add snooze command skeleton

JoelHo committed:
Refactor date handling

Refactor date handling in Duke to accept more human friendly date

Extract magic strings into DukeString

JoelHo committed:
Fix styling issues

JoelHo committed:
Add snooze functionality

Add snooze functionality to Duke

Similar syntax to deadline/event (/at /by /to)

Usage:
snooze idx /at/by date1 (/to date2)

JoelHo committed:
Extract cases from parseInput to methods

Extract long cases from parseInput to individual methods to better fit code quality standards

JoelHo committed:
Merge branch 'branch-B-Snooze'

JoelHo committed:
Add graceful exit to GUI

Add graceful exit to GUI via bye command

JoelHo committed:
Set theme jekyll-theme-slate

JoelHo committed:
Update Duke to use different avatars

Lemon photo by Mockup Graphics on Unsplash

JoelHo committed:
Update title, Add Ui screenshot

Update GUI title to say 'Duke'
Add Ui screenshot for user guide

JoelHo committed:
Update User guide
